src/worker/worker.ml

```diff
@@ -3,6 +3,105 @@ open Std
+  x##send Js.null;
+  match x##.status with
+  | 200 ->
+      Js.Opt.case
+        (File.CoerceTo.arrayBuffer x##.response)
+        (fun () ->
+          Firebug.console##log (Js.string "Failed to receive file");
+          None)
+        (fun b -> Some (Typed_array.String.of_arrayBuffer b))
+  | _ -> None
+
+type signature = Ocaml_typing.Types.signature_item list
+type flags = Ocaml_typing.Cmi_format.pers_flags list
+type header = string * signature
+type crcs = (string * Digest.t option) list 
+
+(** The following two functions are taken from cmi_format.ml in
+    the compiler, but changed to work on bytes rather than input
+    channels *)
+let input_cmi str =
+  let offset = 0 in
+  let (name, sign) = (Marshal.from_bytes str offset : header) in
+  let offset = offset + Marshal.total_size str offset in
+  let crcs = (Marshal.from_bytes str offset : crcs) in
+  let offset = offset + Marshal.total_size str offset in
+  let flags = (Marshal.from_bytes str offset : flags) in
+  {
+    Ocaml_typing.Cmi_format.cmi_name = name;
+    cmi_sign = sign;
+    cmi_crcs = crcs;
+    cmi_flags = flags;
+  }
+
+let read_cmi filename str =
+  let magic = Ocaml_utils.Config.cmi_magic_number in
+  let magic_len = String.length magic in
+  let buffer = Bytes.sub str 0 magic_len in
+  (if buffer <> Bytes.of_string magic then
+   let pre_len = String.length magic - 3 in
+   if
+     Bytes.sub buffer 0 pre_len
+     = Bytes.of_string @@ String.sub magic ~pos:0 ~len:pre_len
+   then
+     let msg =
+       if buffer < Bytes.of_string magic then "an older"
+       else "a newer"
+     in
+     raise (Ocaml_typing.Magic_numbers.Cmi.Error (Wrong_version_interface (filename, msg)))
+   else raise (Ocaml_typing.Magic_numbers.Cmi.Error (Not_an_interface filename)));
+  input_cmi (Bytes.sub str magic_len (Bytes.length str - magic_len))
+
+let memoize f =
+  let memo = Hashtbl.create 10 in
+  fun x ->
+    let open Js_of_ocaml in
+    match Hashtbl.find_opt memo x with
+    | Some x -> x
+    | None ->
+      let result = f x in
+      Hashtbl.replace memo x result;
+      result
+
+let init cmi_urls =
+  let cmi_files =
+    List.map
+      ~f:(fun cmi -> (Filename.basename cmi |> Filename.chop_extension, cmi))
+      cmi_urls
+  in
+  let open Ocaml_typing.Persistent_env.Persistent_signature in
+  let old_loader = !load in
+  let fetch = memoize
+    (fun unit_name ->
+      let open Option.Infix in 
+      List.assoc_opt (String.uncapitalize_ascii unit_name) cmi_files >>= sync_get >>| Bytes.of_string)
+  in
+  let new_load ~unit_name =
+    match fetch unit_name with
+    | Some x ->
+      Some {
+        filename = Sys.executable_name;
+        cmi = read_cmi unit_name x;
+      }
+    | _ -> old_loader ~unit_name
+  in
+  load := new_load
+
+let init { Protocol.static_cmis; cmi_urls } =
+  List.iter static_cmis ~f:(fun ( path, content) ->
+    let name = Filename.(concat "/static/stdlib" (basename path)) in
+    Js_of_ocaml.Sys_js.create_file ~name ~content);
```

Comment:
No - this PR just adds a new way to get the CMIs _in addition_ to the static way. So you get to choose whether you include them statically in your code (the old way) or download them on-demand from the server (the new way). If you choose the new way it does indeed only download them if you ask for them. What's required for the new way though is the list of available CMIs and their urls, so we do need to know ahead of time which CMIs exist.

Reply:
Oh right, I see now!

Anyway, I think the on-demand loading code could be simplified a lot if you used `Sys_js.mount` instead of providing a custom `load` function. This should remove the need for your custom `input_cmi` / `read_cmi`. 
Also this "spoof" filesystem should also work out-of-the-box for any compiler artifact such as `cmt`s, and not only `cmi`s.

But I didn't try it so I might be missing some important detail.

Reply:
I recall trying this and it didn't work - I think it was that directory listing failed maybe? It was a while ago so I have forgotten the details.

Reply:
Yes, that was it - looking up CMIs in [Persistent_env](https://github.com/ocaml/merlin/blob/master/src/ocaml/typing/persistent_env.ml#L41) - however that goes through a different path in merlin than in the compiler (where all this was prototyped), so it might be worth checking to see if mounting does work.

Reply:
OK I had a quick go with `Sys_js.mount`, and it doesn't work. It does produce completions if I 'manually' open the file (ie, `open_in "/static/stdlib.cmi"` at some point after I've done the mount) - then I get the completions I'd expect for that module. But if I don't do that it doesn't do anything. I expect this is something to do with listing the directory, which is something that `Sys_js.mount` doesn't handle.

Reply:
Could you push your attempt anyway ? Here or in another branch ?
I am in London right now, so I might not found time before next week, but I would like to have a look.

Meanwhile, I think it is important to fix the OCaml 5 support, so could you split the changes in two: the fix with the correct cmis for OCaml 5, and the improvement of async loading of the cmts ? That would be very helpful !

Reply:
Sorry, forgot to update on this - I pushed the branch here: https://github.com/jonludlam/merlin-js/tree/psuedofs-test